back/app/controllers/web_api/v1/participation_contexts_controller.rb

```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
+
+class WebApi::V1::ParticipationContextsController < ::ApplicationController
+  before_action :set_context_no_clash, only: %i[survey_results submission_count delete_inputs]
+
+  def survey_results
+    results = SurveyResultsGeneratorService.new(@context).generate_results
+    render json: results
+  end
+
+  def submission_count
+    count = SurveyResultsGeneratorService.new(@context).generate_submission_count
+    render json: count
+  end
+
+  def delete_inputs
+    sidefx.before_delete_inputs @context, current_user
+    ActiveRecord::Base.transaction do
+      @context.ideas.each(&:destroy!)
+    end
+    sidefx.before_delete_inputs @context, current_user
+    head :ok
+  end
+
+  private
+
+  def set_context_no_clash
```

Comment:
As you know I am not a fan of using `before_action` to set state in a controller. In methods where the instance variable is used, it is not clear whether it is set. I prefer using an accessor with lazy initialization instead. It is a different pattern, but it is clear then that an object is available when you need it, instead of relying on a `before_action`.
The fact that the subclasses do not call `set_context` for all actions is another reason not to use `before_action`. Some actions, like `create` set the context without calling `set_context`.
And then I have a third reason: most of the time the state is not needed outside the actions, so why would we use an instance variable? Even if actions call methods where the state is needed, they can pass the state as an argument. The code will be much cleaner, because every method uses the objects given to it, rather than relying on state set outside the methods.

That said, I agree with you that using a superclass is not ideal:
* The class is not abstract, as there is no abstract method in it.`set_context` should be the abstract method.
* Often abstract controller classes make things more complicated. You see it here with `set_context_no_clash` (it is unclear what the "no_clash" part means) and with the other before_action `set_context`. Using both names is confusing.
* Using `@context` instead of `@phase` and `@project` in the subclasses makes things more fuzzy. The original names are very clear. `context` is less clear.
* Other developers may think that the superclass can be used as-is. That is a reason not to use controller class hierarchies.
* If you like to extract common code, you could consider using a mixin, but even that comes with its difficulties.

Actually, I think it is not worthwhile having a common superclass. The code will be clearer without the superclass.

Reply:
Good catch! This was a copy-paste mistake from another policy.

Reply:
I completely agree. I'll get rid of the superclass that was introduced.